### PR TITLE
Support usernameless authentication

### PR DESCRIPTION
**Pull request by ottokruse**
*Issue #, if available:* N/A

*Description of changes:* Support usernameless authentication with passkey, i.e. signing in without having the user type in their username (the WebAuthn `userHandle` is used to determine the username)

EXCUSE THE BIG PR

There are many changes, but most notably:

- The usernameless auth flow uses a public API endpoint to get a sign-in challenge
- These sign-in challenges are stored in DynamoDB so that we can verify it was us who created them, and provide replay protection
- Because the API has a public endpoint now, I added WAF with a rate limit rule, to prevent misuse. However to support WAF I had to add REST API Gateway instead of HTTP API Gateway. Having REST in place now, I also added request validation (supported by REST not by HTTP API)
- The UI slightly overhauled to offer the "Sign in with passkey" button, that does not require entry of a username

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cdk/custom-auth/fido2-challenge-api.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  "Strict-Transport-Security": "max-age=31536000; includeSubdomains; preload",
+  "Content-Type": "application/json",
+  "Cache-Control": "no-store",
+};
+
+const _handler: APIGatewayProxyHandler = async (event) => {
+  logger.debug(JSON.stringify(event, null, 2));
+  logger.info("FIDO2 challenge API invocation:", event.path);
+  try {
+    if (event.path === "/sign-in-challenge") {
```
Comment (EricBorland) [inline cdk/custom-auth/fido2-challenge-api.ts:37]: Why this check if you already have this handler added to the "sign-in-challenge" resource. I think this will always be true since the 404 is already handled by APIGW if the resource does not exist, or am I missing something here?

File: cdk/lib/cognito-passwordless.ts [hunk 2]
```diff
@@ -639,20 +792,222 @@ export class Passwordless extends Construct {
...
+      // sign-in-challenge
+      const signInChallenge =
+        this.fido2Api.root.addResource("sign-in-challenge");
+      signInChallenge.addCorsPreflight(defaultCorsOptionsWithoutAuth);
+      signInChallenge.addMethod(
+        "POST",
+        new cdk.aws_apigateway.LambdaIntegration(this.fido2challengeFn),
+        {
+          authorizer: undefined, // public API
+        }
```
Comment (EricBorland) [inline cdk/lib/cognito-passwordless.ts:964]: Why explicitly indicate this? isn't that the default?

File: client/react/hooks.tsx [hunk 3]
```diff
@@ -485,7 +485,7 @@ function _usePasswordless() {
         tokensCb: (tokens) => storeTokens(tokens).then(() => setTokens(tokens)),
       });
       signinIn.signedIn.catch(setLastError);
-      return signinIn.signedIn;
+      return signinIn;
```
Comment (EricBorland) [inline client/react/hooks.tsx:488]: Why this change? you added then the `.signedIn` on all other parts of code calling these functions, didn't you?

**Review by EricBorland — APPROVED** [review]

File: cdk/custom-auth/fido2-challenge-api.ts (hunk 1 above)
Comment (ottokruse) [reply cdk/custom-auth/fido2-challenge-api.ts:37]: Yes that's true. I guess this is a case of "doubly stitched holds better". You think it's confusing?

File: cdk/lib/cognito-passwordless.ts (hunk 2 above)
Comment (ottokruse) [reply cdk/lib/cognito-passwordless.ts:964]: Just wanted to be super explicit about it in the source, because it's so important

File: client/react/hooks.tsx (hunk 3 above)
Comment (ottokruse) [reply client/react/hooks.tsx:488]: I saw it wasn't consistent, other "signInWith..." methods returned the `signingIn` object, which holds not just the `signedIn` promise but also the `abort` function which may be useful to the caller––so I wanted to make it always available in the response. So, made it consistent and return the `signingIn` object always. If that makes sense?
